File: app/src/main/java/forpdateam/ru/forpda/views/CodeEditor.java
package forpdateam.ru.forpda.views;

/**
 * Created by radiationx on 27.07.17.
 */

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v7.widget.AppCompatEditText;
import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.ReplacementSpan;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ScrollView;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import forpdateam.ru.forpda.utils.Html;

/*
* ORIGINAL: https://github.com/markusfisch/CodeEditor/blob/master/app/src/main/java/de/markusfisch/android/CodeEditor/widget/CodeEditor.java
* */
public class CodeEditor extends AppCompatEditText {
    private static class ForumCodes {
        private static final Pattern ELEMENT = Pattern.compile("(\\[(?:\\/)?((?:attachment|background|nomergetime|mergetime|snapback|numlist|spoiler|offtop|center|color|right|quote|code|font|hide|left|list|size|img|sub|sup|cur|url|b|i|u|s|\\*)))=?([^\\]\\[]+?)?(\\])", Pattern.CASE_INSENSITIVE);
        private static final Pattern ATTRIBUTE = Pattern.compile("(name|date|post)?=?(\"[^\\\"]*?\"|[\\w\\d]+)", Pattern.CASE_INSENSITIVE);
    }

    private final Handler updateHandler = new Handler();
    private final Runnable updateRunnable = () -> {
        Editable e = getText();
        highlightWithoutChange(e);
    };

    private int updateDelay = 500;
    private boolean modified = true;
    private int colorTag;
    private int colorAttrName;
    private int colorAttrValue;
    private ScrollView scrollView;

    private Runnable scrollerTask;
    private int initialPosition;
    private int newCheck = 100;

    public void attachToScrollView(ScrollView sv) {
        scrollView = sv;

        scrollView.addOnLayoutChangeListener((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
            int heightWas = oldBottom - oldTop;
            if (v.getHeight() != heightWas) {
                smartUpdateHighlighting();
            }
        });

        scrollView.setOnTouchListener((v, event) -> {
            if (event.getAction() == MotionEvent.ACTION_UP) {
                initialPosition = scrollView.getScrollY();
                scrollView.postDelayed(scrollerTask, newCheck);
            }
            return false;
        });

        scrollerTask = () -> {
            int newPosition = scrollView.getScrollY();
            if (initialPosition - newPosition == 0) {
                smartUpdateHighlighting();
            } else {
                initialPosition = scrollView.getScrollY();
                scrollView.postDelayed(scrollerTask, newCheck);
            }
        };

    }

    public CodeEditor(Context context) {
        super(context);
        init();
    }

    public CodeEditor(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }


    public void setUpdateDelay(int ms) {
        updateDelay = ms;
    }


    public void updateHighlighting() {
        highlightWithoutChange(getText());
    }

    private void init() {
        //setHorizontallyScrolling(true);

        addTextChangedListener(new TextWatcher() {
            @Override
            public void onTextChanged(
                    CharSequence s,
                    int start,
                    int before,
                    int count) {
            }

            @Override
            public void beforeTextChanged(
                    CharSequence s,
                    int start,
                    int count,
                    int after) {
            }

            @Override
            public void afterTextChanged(Editable e) {
                smartUpdateHighlighting();
            }
        });

        setSyntaxColors();
        setUpdateDelay(500);
        updateHighlighting();
    }

    private void smartUpdateHighlighting() {
        cancelUpdate();
        if (!modified) {
            return;
        }
        updateHandler.postDelayed(updateRunnable, updateDelay);
    }

    private void setSyntaxColors() {
        colorTag = Color.parseColor("#446FBD");
        colorAttrName = Color.parseColor("#6D8600");
        colorAttrValue = Color.parseColor("#e88501");
    }

    private void cancelUpdate() {
        updateHandler.removeCallbacks(updateRunnable);
    }

    private void highlightWithoutChange(Editable e) {
        modified = false;
        highlight(e);
        modified = true;
    }

    private Editable highlight(Editable e) {
        long time = System.currentTimeMillis();
        try {
            // don't use e.clearSpans() because it will
            // remove too much
            clearSpans(e);

            if (e.length() == 0) {
                return e;
            }

            int scrollY = scrollView.getScrollY();
            int scrollViewHeight = scrollView.getHeight();


            int visibleStart = getOffsetForPosition(0, scrollY);
            int visibleEnd = getOffsetForPosition(0, scrollY + scrollViewHeight);
            visibleStart = Math.max(0, visibleStart - 100);
            visibleEnd = Math.min(e.length(), visibleEnd + 100);

            CharSequence hlText = e.subSequence(visibleStart, visibleEnd);
            //visibleStart = 0;
            //visibleEnd = 0;
            for (Matcher m = ForumCodes.ELEMENT.matcher(hlText), attributes = null; m.find(); ) {
                String attrsSrc = m.group(3);
                if (attrsSrc != null) {
                    String tagName = m.group(2);
                    int eg3s = m.start(3);
                    int eg3e = m.end(3);
                    int color = colorAttrValue;

                    if (tagName.equalsIgnoreCase("quote")) {
                        if (attributes == null) {
                            attributes = ForumCodes.ATTRIBUTE.matcher(attrsSrc);
                        } else {
                            attributes = attributes.reset(attrsSrc);
                        }

                        while (attributes.find()) {
                            String attrName = attributes.group(1);
                            if (attrName != null) {
                                int ag1s = attributes.start(1);
                                int ag1e = attributes.end(1);
                                e.setSpan(
                                        new ForegroundColorSpan(colorAttrName),
                                        visibleStart + eg3s + ag1s,
                                        visibleStart + eg3s + ag1e,
                                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                            }

                            String attrValue = attributes.group(2);
                            if (attrValue != null) {
                                int ag2s = attributes.start(2);
                                int ag2e = attributes.end(2);

                                e.setSpan(
                                        new ForegroundColorSpan(color),
                                        visibleStart + eg3s + ag2s,
                                        visibleStart + eg3s + ag2e,
                                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                            }


                        }
                    } else {
                        if (tagName.equalsIgnoreCase("color") || tagName.equalsIgnoreCase("background")) {
                            try {
                                if (attrsSrc.charAt(0) != '#') {
                                    Integer htmlColor = Html.getColorMap().get(attrsSrc);
                                    if (htmlColor != null) {
                                        color = htmlColor;
                                    }
                                } else {
                                    color = Color.parseColor(attrsSrc);
                                }
                            } catch (Exception ignore) {
                            }
                        }
                        e.setSpan(
                                new ForegroundColorSpan(color),
                                visibleStart + eg3s,
                                visibleStart + eg3e,
                                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                }

                int eg1s = m.start(1);
                int eg1e = m.end(1);
                e.setSpan(
                        new ForegroundColorSpan(colorTag),
                        visibleStart + eg1s,
                        visibleStart + eg1e,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

                int eg4s = m.start(4);
                int eg4e = m.end(4);
                e.setSpan(
                        new ForegroundColorSpan(colorTag),
                        visibleStart + eg4s,
                        visibleStart + eg4e,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }

        } catch (IllegalStateException ex) {
            // raised by Matcher.start()/.end() when
            // no successful match has been made what
            // shouldn't ever happen because of find()
        }

        Log.d("CodeEditor", "Time: " + (System.currentTimeMillis() - time));
        return e;
    }

    private static void clearSpans(Editable e) {
        // remove foreground color spans
        {
            ForegroundColorSpan spans[] = e.getSpans(
                    0,
                    e.length(),
                    ForegroundColorSpan.class);

            for (int i = spans.length; i-- > 0; ) {
                e.removeSpan(spans[i]);
            }
        }

        // remove background color spans
        {
            BackgroundColorSpan spans[] = e.getSpans(
                    0,
                    e.length(),
                    BackgroundColorSpan.class);

            for (int i = spans.length; i-- > 0; ) {
                e.removeSpan(spans[i]);
            }
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/views/CodeEditor.kt
package forpdateam.ru.forpda.ui.views

import android.content.Context
import android.graphics.Color
import android.os.Handler
import android.text.Editable
import android.text.Spannable
import android.text.TextWatcher
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.ScrollView
import androidx.appcompat.widget.AppCompatEditText
import forpdateam.ru.forpda.common.Html
import java.util.regex.Matcher
import java.util.regex.Pattern
import kotlin.math.max
import kotlin.math.min

/**
 * Created by radiationx on 27.07.17.
 */

/*
 * ORIGINAL: https://github.com/markusfisch/CodeEditor/blob/master/app/src/main/java/de/markusfisch/android/CodeEditor/widget/CodeEditor.java
 * */
class CodeEditor : AppCompatEditText {
    private object ForumCodes {
        val ELEMENT: Pattern = Pattern.compile(
            "(\\[(?:\\/)?((?:attachment|background|nomergetime|mergetime|snapback|numlist|spoiler|offtop|center|color|right|quote|code|font|hide|left|list|size|img|sub|sup|cur|url|b|i|u|s|\\*)))=?\\s?([^\\]\\[]+?)?(\\])",
            Pattern.CASE_INSENSITIVE
        )
        val ATTRIBUTE: Pattern = Pattern.compile(
            "(name|date|post)?=?([\\s\\S]+?)\\s?(?=(?:name|date|post)=|\\z)",
            Pattern.CASE_INSENSITIVE
        )
    }

    private val updateHandler = Handler()
    private val updateRunnable = Runnable {
        val e = text
        highlightWithoutChange(e!!)
    }

    private var updateDelay = 500
    private var modified = true
    private var colorTag = 0
    private var colorAttrName = 0
    private var colorAttrValue = 0
    private var scrollView: ScrollView? = null

    private var scrollerTask: Runnable? = null
    private var initialPosition = 0
    private val newCheck = 100

    fun attachToScrollView(sv: ScrollView?) {
        scrollView = sv

        scrollView!!.addOnLayoutChangeListener { v: View, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int ->
            val heightWas = oldBottom - oldTop
            if (v.height != heightWas) {
                smartUpdateHighlighting()
            }
        }

        scrollView!!.setOnTouchListener { v: View?, event: MotionEvent ->
            if (event.action == MotionEvent.ACTION_UP) {
                initialPosition = scrollView!!.scrollY
                scrollView!!.postDelayed(scrollerTask, newCheck.toLong())
            }
            false
        }

        scrollerTask = Runnable {
            val newPosition = scrollView!!.scrollY
            if (initialPosition - newPosition == 0) {
                smartUpdateHighlighting()
            } else {
                initialPosition = scrollView!!.scrollY
                scrollView!!.postDelayed(scrollerTask, newCheck.toLong())
            }
        }
    }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }


    fun setUpdateDelay(ms: Int) {
        updateDelay = ms
    }


    fun updateHighlighting() {
        highlightWithoutChange(text!!)
    }

    private fun init() {
        //setHorizontallyScrolling(true);

        addTextChangedListener(object : TextWatcher {
            override fun onTextChanged(
                s: CharSequence,
                start: Int,
                before: Int,
                count: Int
            ) {
            }

            override fun beforeTextChanged(
                s: CharSequence,
                start: Int,
                count: Int,
                after: Int
            ) {
            }

            override fun afterTextChanged(e: Editable) {
                smartUpdateHighlighting()
            }
        })

        setSyntaxColors()
        setUpdateDelay(500)
        updateHighlighting()
    }

    private fun smartUpdateHighlighting() {
        cancelUpdate()
        if (!modified) {
            return
        }
        updateHandler.postDelayed(updateRunnable, updateDelay.toLong())
    }

    private fun setSyntaxColors() {
        colorTag = Color.parseColor("#446FBD")
        colorAttrName = Color.parseColor("#6D8600")
        colorAttrValue = Color.parseColor("#e88501")
    }

    private fun cancelUpdate() {
        updateHandler.removeCallbacks(updateRunnable)
    }

    private fun highlightWithoutChange(e: Editable) {
        modified = false
        highlight(e)
        modified = true
    }

    private fun highlight(e: Editable): Editable? {
        val time = System.currentTimeMillis()
        try {
            // don't use e.clearSpans() because it will
            // remove too much
            clearSpans(e)

            if (e.length == 0) {
                return e
            }

            val scrollY = scrollView!!.scrollY
            val scrollViewHeight = scrollView!!.height


            var visibleStart = getOffsetForPosition(0f, scrollY.toFloat())
            var visibleEnd = getOffsetForPosition(0f, (scrollY + scrollViewHeight).toFloat())
            visibleStart = max(0.0, (visibleStart - 100).toDouble()).toInt()
            visibleEnd = min(e.length.toDouble(), (visibleEnd + 100).toDouble()).toInt()

            val hlText = e.subSequence(visibleStart, visibleEnd)
            //visibleStart = 0;
            //visibleEnd = 0;
            val m = ForumCodes.ELEMENT.matcher(hlText)
            var attributes: Matcher? = null
            while (m.find()) {
                val attrsSrc = m.group(3)
                if (attrsSrc != null) {
                    val tagName = m.group(2)
                    val eg3s = m.start(3)
                    val eg3e = m.end(3)
                    var color = colorAttrValue

                    if (tagName.equals("quote", ignoreCase = true)) {
                        attributes = if (attributes == null) {
                            ForumCodes.ATTRIBUTE.matcher(attrsSrc)
                        } else {
                            attributes.reset(attrsSrc)
                        }

                        while (attributes.find()) {
                            val attrName = attributes.group(1)
                            if (attrName != null) {
                                val ag1s = attributes.start(1)
                                val ag1e = attributes.end(1)
                                e.setSpan(
                                    ForegroundColorSpan(colorAttrName),
                                    visibleStart + eg3s + ag1s,
                                    visibleStart + eg3s + ag1e,
                                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                                )
                            }

                            val attrValue = attributes.group(2)
                            if (attrValue != null) {
                                val ag2s = attributes.start(2)
                                val ag2e = attributes.end(2)

                                e.setSpan(
                                    ForegroundColorSpan(color),
                                    visibleStart + eg3s + ag2s,
                                    visibleStart + eg3s + ag2e,
                                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                                )
                            }
                        }
                    } else {
                        if (tagName.equals(
                                "color",
                                ignoreCase = true
                            ) || tagName.equals("background", ignoreCase = true)
                        ) {
                            try {
                                if (attrsSrc[0] != '#') {
                                    val htmlColor = Html.getColorMap()[attrsSrc]
                                    if (htmlColor != null) {
                                        color = htmlColor
                                    }
                                } else {
                                    color = Color.parseColor(attrsSrc)
                                }
                            } catch (ignore: Exception) {
                            }
                        }
                        e.setSpan(
                            ForegroundColorSpan(color),
                            visibleStart + eg3s,
                            visibleStart + eg3e,
                            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                        )
                    }
                }

                val eg1s = m.start(1)
                val eg1e = m.end(1)
                e.setSpan(
                    ForegroundColorSpan(colorTag),
                    visibleStart + eg1s,
                    visibleStart + eg1e,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                )

                val eg4s = m.start(4)
                val eg4e = m.end(4)
                e.setSpan(
                    ForegroundColorSpan(colorTag),
                    visibleStart + eg4s,
                    visibleStart + eg4e,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                )
            }
        } catch (ex: IllegalStateException) {
            // raised by Matcher.start()/.end() when
            // no successful match has been made what
            // shouldn't ever happen because of find()
        }

        Log.d("CodeEditor", "Time: " + (System.currentTimeMillis() - time))
        return e
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        cancelUpdate()
    }

    companion object {
        private fun clearSpans(e: Editable) {
            // remove foreground color spans
            run {
                val spans = e.getSpans(
                    0,
                    e.length,
                    ForegroundColorSpan::class.java
                )
                var i = spans.size
                while (i-- > 0) {
                    e.removeSpan(spans[i])
                }
            }

            // remove background color spans
            run {
                val spans = e.getSpans(
                    0,
                    e.length,
                    BackgroundColorSpan::class.java
                )
                var i = spans.size
                while (i-- > 0) {
                    e.removeSpan(spans[i])
                }
            }
        }
    }
}